Safari 16 cannot play WebM Opus from blob URL (breaks Construct 3)

jyavenard opened the pull request:
#### a7b863a49945946c913e6e194ec047da844094a4
<pre>
Safari 16 cannot play WebM Opus from blob URL (breaks Construct 3)
<a href="https://bugs.webkit.org/show_bug.cgi?id=245428">https://bugs.webkit.org/show_bug.cgi?id=245428</a>
<a href="https://rdar.apple.com/100205028">rdar://100205028</a>

Reviewed by Youenn Fablet.

The Blob&apos;s content-type wasn&apos;t passed to the MediaPlayer preventing playback to succeed.
Similarly, when a Blob&apos;s URL was provided, an empty content-type was used.

We add a BlobRegistry::blobType method, which similar to blobSize will fetch the content-type
from the BlobData in the network process.

Added tests.

* LayoutTests/http/tests/media/video-src-mp4-blob-expected.txt: Added.
* LayoutTests/http/tests/media/video-src-mp4-blob.html: Added.
* LayoutTests/http/tests/media/video-src-webm-blob-expected.txt: Added.
* LayoutTests/http/tests/media/video-src-webm-blob.html: Added.
* LayoutTests/http/tests/media/video-srcobject-webm-blob-expected.txt: Added.
* LayoutTests/http/tests/media/video-srcobject-webm-blob.html: Added.
* LayoutTests/http/tests/media/white.webm: File was a mkv with a h264 video track. Re-encoded it to use a vp9 track instead.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::blobType):
* Source/WebCore/fileapi/ThreadableBlobRegistry.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::loadResource):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/network/BlobRegistry.h:
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::BlobRegistryImpl::blobType):
* Source/WebCore/platform/network/BlobRegistryImpl.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::blobType):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessPlatformStrategies.cpp:
(WebKit::NetworkProcessPlatformStrategies::createBlobRegistry):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.cpp:
(WebKit::BlobRegistryProxy::blobType):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:

Canonical link: <a href="https://commits.webkit.org/276136@main">https://commits.webkit.org/276136@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3cc2c17fd90acddf60281f1fbd86f39553737cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22875 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46470 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39919 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46130 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20279 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36177 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44400 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19907 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37750 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17173 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17430 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38829 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1880 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39997 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39129 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48030 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18846 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15426 "Found 1 new test failure: http/tests/navigation/parsed-iframe-dynamic-form-back-entry.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42986 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20242 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38021 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9753 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20443 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19866 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
```diff
@@ -175,6 +175,19 @@ void ThreadableBlobRegistry::registerInternalBlobURLForSlice(const URL& newURL,
+
+    String result;
+    callOnMainThreadAndWait([url = url.isolatedCopy(), &result] {
+        result = blobRegistry().blobType(url);
```
We probably want to isolate copy result, this should hopefully not create any memory allocation.
Or simply return early if not main thread.

youennf commented on Source/WebCore/platform/network/BlobRegistryImpl.cpp:
```diff
@@ -301,6 +301,16 @@ BlobData* BlobRegistryImpl::getBlobDataFromURL(const URL& url, const std::option
+String BlobRegistryImpl::blobType(const URL& url)
+{
+    ASSERT(isMainThread());
+    BlobData* data = getBlobDataFromURL(url);
```
```suggestion
    RefPtr data = getBlobDataFromURL(url);
```

youennf reviewed (approved):
These are all good tests we should upstream to WPT.
It would be better to write them as testharness.js tests

jyavenard committed:
Safari 16 cannot play WebM Opus from blob URL (breaks Construct 3)
https://bugs.webkit.org/show_bug.cgi?id=245428
rdar://100205028

Reviewed by Youenn Fablet.

The Blob's content-type wasn't passed to the MediaPlayer preventing playback to succeed.
Similarly, when a Blob's URL was provided, an empty content-type was used.

We add a BlobRegistry::blobType method, which similar to blobSize will fetch the content-type
from the BlobData in the network process.

Added tests.

* LayoutTests/http/tests/media/video-src-mp4-blob-expected.txt: Added.
* LayoutTests/http/tests/media/video-src-mp4-blob.html: Added.
* LayoutTests/http/tests/media/video-src-webm-blob-expected.txt: Added.
* LayoutTests/http/tests/media/video-src-webm-blob.html: Added.
* LayoutTests/http/tests/media/video-srcobject-webm-blob-expected.txt: Added.
* LayoutTests/http/tests/media/video-srcobject-webm-blob.html: Added.
* LayoutTests/http/tests/media/white.webm: File was a mkv with a h264 video track. Re-encoded it to use a vp9 track instead.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/fileapi/ThreadableBlobRegistry.cpp:
(WebCore::ThreadableBlobRegistry::blobType):
* Source/WebCore/fileapi/ThreadableBlobRegistry.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::loadResource):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/network/BlobRegistry.h:
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::BlobRegistryImpl::blobType):
* Source/WebCore/platform/network/BlobRegistryImpl.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::blobType):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessPlatformStrategies.cpp:
(WebKit::NetworkProcessPlatformStrategies::createBlobRegistry):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.cpp:
(WebKit::BlobRegistryProxy::blobType):
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:

Canonical link: https://commits.webkit.org/276136@main
